pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java

```diff
@@ -137,42 +137,39 @@ public boolean equals(Object obj) {
     }
 
     public void publishMessage(long producerId, long sequenceId, ByteBuf headersAndPayload, long batchSize) {
-        beforePublish(producerId, sequenceId, headersAndPayload, batchSize);
-        publishMessageToTopic(headersAndPayload, sequenceId, batchSize);
+        if (checkAndStartPublish(producerId, sequenceId, headersAndPayload, batchSize)) {
+            publishMessageToTopic(headersAndPayload, sequenceId, batchSize);
+        }
     }
 
     public void publishMessage(long producerId, long lowestSequenceId, long highestSequenceId,
-           ByteBuf headersAndPayload, long batchSize) {
+            ByteBuf headersAndPayload, long batchSize) {
         if (lowestSequenceId > highestSequenceId) {
-            cnx.ctx().channel().eventLoop().execute(() -> {
-                cnx.ctx().writeAndFlush(Commands.newSendError(producerId, highestSequenceId, ServerError.MetadataError,
-                        "Invalid lowest or highest sequence id"));
+            execute(() -> {
+                sendError(producerId, highestSequenceId, ServerError.MetadataError, "Invalid lowest or highest sequence id");
                 cnx.completedSendOperation(isNonPersistentTopic, headersAndPayload.readableBytes());
             });
-            return;
         }
-        beforePublish(producerId, highestSequenceId, headersAndPayload, batchSize);
```

Comment:
beforePublish makes some checks.
publishMessageToTopic shouldn't be called if those checks fail.